### PR TITLE
[SPARK-21196] Split codegen info of query plan into sequence

### PR DESCRIPTION
**Pull request by gengliangwang**
codegen info of query plan can be very long. 
In debugging console / web page, it would be more readable if the subtrees and corresponding codegen are split into sequence.

Example:

```java
codegenStringSeq(sql("select 1").queryExecution.executedPlan)
```
The example will return Seq[(String, String)] of length 1, containing the subtree as string and the corresponding generated code.

The subtree as string:

> (*Project [1 AS 1#0]
> +- Scan OneRowRelation[]

The generated code:
```java
/* 001 */ public Object generate(Object[] references) {
/* 002 */   return new GeneratedIterator(references);
/* 003 */ }
/* 004 */
/* 005 */ final class GeneratedIterator extends org.apache.spark.sql.execution.BufferedRowIterator {
/* 006 */   private Object[] references;
/* 007 */   private scala.collection.Iterator[] inputs;
/* 008 */   private scala.collection.Iterator inputadapter_input;
/* 009 */   private UnsafeRow project_result;
/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder project_holder;
/* 011 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter project_rowWriter;
/* 012 */
/* 013 */   public GeneratedIterator(Object[] references) {
/* 014 */     this.references = references;
/* 015 */   }
/* 016 */
/* 017 */   public void init(int index, scala.collection.Iterator[] inputs) {
/* 018 */     partitionIndex = index;
/* 019 */     this.inputs = inputs;
/* 020 */     inputadapter_input = inputs[0];
/* 021 */     project_result = new UnsafeRow(1);
/* 022 */     project_holder = new org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder(project_result, 0);
/* 023 */     project_rowWriter = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(project_holder, 1);
/* 024 */
/* 025 */   }
/* 026 */
/* 027 */   protected void processNext() throws java.io.IOException {
/* 028 */     while (inputadapter_input.hasNext() && !stopEarly()) {
/* 029 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 030 */       project_rowWriter.write(0, 1);
/* 031 */       append(project_result);
/* 032 */       if (shouldStop()) return;
/* 033 */     }
/* 034 */   }
/* 035 */
/* 036 */ }
```
## What changes were proposed in this pull request?
add method codegenToSeq: split codegen info of query plan into sequence

## How was this patch tested?
unit test

@cloud-fan @gatorsmile 
Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 1]
```diff
@@ -50,23 +50,38 @@ package object debug {
...
+      output += s"== Subtree ${i + 1} / ${codegenSeq.size} ==\n"
+      output += subtree
+      output += "\nGenerated code:\n"
+      output += s"${code}\n"
+    }
+    output
+  }
+
+  /**
+   * Find WholeStageCodegenExec subtrees in query plan and do codegen for each of them
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: > Outputs the WholeStageCodegenExec subtrees and the codegen in a query plan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala [hunk 2]
```diff
@@ -38,4 +38,11 @@ class DebuggingSuite extends SparkFunSuite with SharedSQLContext {
     assert(res.contains("Subtree 2 / 2"))
     assert(res.contains("Object[]"))
   }
+
+  test("debugCodegenStringSeq") {
+    val res = codegenStringSeq(spark.range(10).groupBy("id").count().queryExecution.executedPlan)
+    assert(res.length == 2)
+    assert(res.seq.forall{case (subtree, code) =>
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala]: `forall{case` -> `forall { case`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala [hunk 3]
```diff
@@ -38,4 +38,11 @@ class DebuggingSuite extends SparkFunSuite with SharedSQLContext {
     assert(res.contains("Subtree 2 / 2"))
     assert(res.contains("Object[]"))
   }
+
+  test("debugCodegenStringSeq") {
+    val res = codegenStringSeq(spark.range(10).groupBy("id").count().queryExecution.executedPlan)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala]: Could you please post the return value here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 4]
```diff
@@ -50,23 +50,38 @@ package object debug {
...
     }
-    var output = s"Found ${codegenSubtrees.size} WholeStageCodegen subtrees.\n"
-    for ((s, i) <- codegenSubtrees.toSeq.zipWithIndex) {
-      output += s"== Subtree ${i + 1} / ${codegenSubtrees.size} ==\n"
-      output += s
-      output += "\nGenerated code:\n"
-      val (_, source) = s.doCodeGen()
-      output += s"${CodeFormatter.format(source)}\n"
+    codegenSubtrees.toSeq.map {
+      subtree =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: Please collapse the above two lines into a single line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 5]
```diff
@@ -50,23 +50,38 @@ package object debug {
     // scalastyle:on println
   }
 
+  /** Generate codegen debug info */
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: Please add the same function description like what you did in `codegenStringSeq `

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 6]
```diff
@@ -245,5 +245,10 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
       println(org.apache.spark.sql.execution.debug.codegenString(executedPlan))
       // scalastyle:on println
     }
+
+    /** @return Sequence of WholeStageCodegen subtrees and corresponding codegen */
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: The same here

**Comment by gengliangwang** [comment]
Example:

```java
codegenStringSeq(sql("select 1").queryExecution.executedPlan)
```
The example will return Seq[(String, String)] of length 1, containing the subtree as string and the corresponding generated code.

The subtree as string:

> (*Project [1 AS 1#0]
> +- Scan OneRowRelation[]

The generated code:
```java
/* 001 */ public Object generate(Object[] references) {
/* 002 */   return new GeneratedIterator(references);
/* 003 */ }
/* 004 */
/* 005 */ final class GeneratedIterator extends org.apache.spark.sql.execution.BufferedRowIterator {
/* 006 */   private Object[] references;
/* 007 */   private scala.collection.Iterator[] inputs;
/* 008 */   private scala.collection.Iterator inputadapter_input;
/* 009 */   private UnsafeRow project_result;
/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder project_holder;
/* 011 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter project_rowWriter;
/* 012 */
/* 013 */   public GeneratedIterator(Object[] references) {
/* 014 */     this.references = references;
/* 015 */   }
/* 016 */
/* 017 */   public void init(int index, scala.collection.Iterator[] inputs) {
/* 018 */     partitionIndex = index;
/* 019 */     this.inputs = inputs;
/* 020 */     inputadapter_input = inputs[0];
/* 021 */     project_result = new UnsafeRow(1);
/* 022 */     project_holder = new org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder(project_result, 0);
/* 023 */     project_rowWriter = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(project_holder, 1);
/* 024 */
/* 025 */   }
/* 026 */
/* 027 */   protected void processNext() throws java.io.IOException {
/* 028 */     while (inputadapter_input.hasNext() && !stopEarly()) {
/* 029 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 030 */       project_rowWriter.write(0, 1);
/* 031 */       append(project_result);
/* 032 */       if (shouldStop()) return;
/* 033 */     }
/* 034 */   }
/* 035 */
/* 036 */ }
```

**Comment by gengliangwang** [comment]
@gatorsmile Thanks, I have revised the code

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @cloud-fan

**Comment by gengliangwang** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Pull request closed without merging**